Add build and send sweep transaction for bitcoin js wallet

matthewjablack opened the pull request:
### Description

This PR adds `buildSweepTransaction` and `sendSweepTransaction` for Bitcoin js wallet enabling the ability to sweep all utxos to a destination of choice. 

This is particularly useful if you wish to remove all funds from a wallet

Note: This can be useful for Ethereum as well. The major use case would be for sweeping all ETH from a wallet (since it's already pretty easy to do this for ERC20's)

### Submission Checklist :pencil:

- [x] Add `buildSweepTransaction
- [x] Add `sendSweepTransaction
- [x] Add tests for `sendSweepTransaction`

matthewjablack committed:
Add build and send sweep transaction for js wallet

harshjv commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.js:
```diff
@@ -92,6 +92,29 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(Walle
+    const { inputs, outputs, change } = await this.getInputsForSweep(_outputs, _feePerByte, fixedInputs)
+
+    if (change) {
+      throw Error('There should not be any change for sweeping transaction')
```
I think you missed a `new` here ;)

monokh commented:
Hey @matthewjablack thanks for the PR here, this is a very cool function
To keep the compatibility of the library uniform, we usually need any new functions to have implementations across ETH and ERC20 too, is this something you can add?

matthewjablack commented:
Hey @monokh we're pretty focused on Bitcoin development at the moment, so unfortunately the ETH side of things would be out of scope of us.

matthewjablack committed:
Adjust input output byte size calculation for fee per byte

matthewjablack committed:
Modify tests to ensure necessary funds when js wallet is sweeped

matthewjablack committed:
Change input and output fee calculation

monokh commented:
@matthewjablack To ensure the library is predictable and fits the purpose of being an abstraction layer between the chains, we really need any functionality implemented to have a suitable interface and be implemented across the supported chains where possible.

Maintaining single chain based functionality adds overhead that the library is not suited towards. For example the current signature for `async sendSweepTransaction (externalChangeAddress, outputs, fee, fixedInputs)` is not compatible with the ETH chain and will need breaking changes when implemented cross chain.

So we ask for new features to be implemented cross chain because it:

- Forces the right interface to be created at the client level
- Ensures that the functionality is indeed able to be abstracted in a usable way
- Ensures that between the chains, the function is uniform and behaves in a predictable way
- Works as expected for the target users of the library (those looking to abstract cross chain)

matthewjablack committed:
Add sendSweepTransaction to Ethereum js wallet

matthewjablack committed:
Fix bitcoin tests for sweep tx

matthewjablack commented:
Hey @monokh good point about `async sendSweepTransaction (externalChangeAddress, outputs, fee, fixedInputs)` not being compatible with the ETH chain. 

I've changed it to `async sendSweepTransaction (externalAddress, fee, outputs, fixedInputs)` and implemented it on the ETH chain

monokh commented on packages/bitcoin-wallet-provider/lib/BitcoinWalletProvider.js:
```diff
@@ -226,6 +236,102 @@ export default superclass => class BitcoinWalletProvider extends superclass {
       .then(({ unusedAddress }) => unusedAddress[key])
   }
 
+  async getInputsForSweep (_targets = [], _feePerByte, fixedInputs = [], numAddressPerCall = 100) {
```
This looks like a lot of duplication from `getInputsForAmount`. Can the logic there be reused?

matthewjablack replied:
DRYed this out by moving logic for `getInputsForSweep` into `getInputsForAmount` without affecting existing functionality of `getInputsForAmount` in https://github.com/liquality/chainabstractionlayer/pull/358/commits/68355ac3b6fc671b8630cd7f44415609f907d633

monokh commented on packages/bitcoin-wallet-provider/lib/BitcoinWalletProvider.js:
```diff
@@ -226,6 +236,102 @@ export default superclass => class BitcoinWalletProvider extends superclass {
+
+      const outputBalance = _targets.reduce((a, b) => a + (b['value'] || 0), 0)
+
+      const amountToSend = utxoBalance - (feePerByte * (((_targets.length + 1) * 39) + (utxos.length * 153))) // todo better calculation
```
I guess this would depend on the type of tx. Can https://github.com/bitcoinjs/coinselect be used here in any way?

matthewjablack replied:
Tests passed with `legacy`, `p2sh-segwit` and `bech32`

We know that all utxos spent through this method, will be a p2pkh or p2wpkh txs (since we're spending from the wallet and not from p2sh or p2wsh txs) 

Looking at: https://github.com/bitcoinjs/coinselect/blob/master/utils.js#L3 we can see that

```
var TX_INPUT_BASE = 32 + 4 + 1 + 4
var TX_INPUT_PUBKEYHASH = 107
var TX_OUTPUT_BASE = 8 + 1
var TX_OUTPUT_PUBKEYHASH = 25
```

So a pkh input is (32 + 4 + 1 + 4) + 107 = 148
and a pkh output is (8 + 1) + 25 = 34

I added a bit of padding (5 bytes each) to arrive at 39 and 153 respectively

monokh replied:
This looks good. Unfortunately fees in `coinselect` are pre segwit, so likely not much we can do here for now: https://github.com/bitcoinjs/coinselect/issues/53

monokh commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.js:
```diff
@@ -92,6 +92,29 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(Walle
+      throw Error('There should not be any change for sweeping transaction')
+    }
+
+    _outputs.forEach((output, i) => {
```
What is going on here with the `_outputs` and `fixedInputs` ?
The common set of function across chains would be to send all value in the wallet to a single address.

matthewjablack replied:
Having `_outputs` and `fixedInputs` allows you to spend all utxos in a wallet, and send a specific amount to an address, while sending the change to another address. 

Useful if you want to pay someone and sweep your wallet at the same time.

monokh replied:
Ok, i see what you are going for here. However i think the functions should not cater for complex features, unless it can be abstracted cross chain or is required for a specific need.

monokh replied:
we want to reduce our maintenance to what is necessary for common usage

monokh replied:
changing that should make this function as well as `getInputsForAmount` simpler

matthewjablack replied:
There are actually a couple of use cases where this is quite important. 

If you want to allow users to select specific utxos and spend them or send funds to a particular location (similar to Wasabi) having defined inputs and outputs is necessary. 

Another case is for DLC's. Every DLC created, must commit to a specific utxo for the funding tx. If the user wants to create multiple DLC's, they must split up the utxos beforehand. To do this as efficiently as possible, it's important to be able to specify a particular utxo, create one output to be used for a specific DLC, and the other one be change.

monokh commented on packages/client/lib/Chain.js:
```diff
@@ -187,6 +187,30 @@ export default class Chain {
+   * @param {string[]} fixedInputs - txid, vout, address, value, satoshis, derivationPath
+   * @return {Promise<string>} Resolves with a signed transaction object.
+   */
+  async buildSweepTransaction (externalAddress, fee, outputs, fixedInputs) {
```
If this is not something that is implemented cross chain, it shouldn't be included at the client level. In special situations where you need to access something on a specific chain, you can do `client.getMethod('buildSweepTransaction')` or access it directly on the provider object.

matthewjablack replied:
Good point, updated in https://github.com/liquality/chainabstractionlayer/pull/358/commits/3100c572312cdd9a3510e223d7ec76ca921a060e

monokh commented on packages/client/lib/Chain.js:
```diff
@@ -187,6 +187,30 @@ export default class Chain {
+   * @param {string[]} fixedInputs - txid, vout, address, value, satoshis, derivationPath
+   * @return {Promise<Transaction>} Resolves with a signed transaction.
+   */
+  async sendSweepTransaction (externalAddress, fee, outputs, fixedInputs) {
```
On the client, signatures should not include anything that we are not able to abstract across chains. Should be `sendSweepTransaction (address, fee)`

matthewjablack replied:
Good point, updated in https://github.com/liquality/chainabstractionlayer/pull/358/commits/3100c572312cdd9a3510e223d7ec76ca921a060e

monokh commented on packages/ethereum-js-wallet-provider/lib/EthereumJsWalletProvider.js:
```diff
@@ -105,6 +106,21 @@ export default class EthereumJsWalletProvider extends Provider {
+    const balance = await this.getMethod('getBalance')(addresses)
+
+    const [ gasPrice ] = await Promise.all([
+      _gasPrice ? Promise.resolve(_gasPrice) : this.getMethod('getGasPrice')()
```
I think this comes back as wei? Later on in `sendTransaction` we pass this as the fee parameter (`gasPrice`) and on that level it expects it in gwei.

matthewjablack replied:
So `getGasPrice` function returns the value in `gwei` (tested locally with ganache, idk if it it's different with parity/geth/infura)

So I'm pretty sure gasPrice in `gwei` is being passed around (otherwise there would be issues here: https://github.com/liquality/chainabstractionlayer/blob/dev/packages/ethereum-js-wallet-provider/lib/EthereumJsWalletProvider.js#L96 )

monokh reviewed (changes requested):
Please also include the ERC20 implementation. I believe it would be in `EthereumErc20Provider`

matthewjablack committed:
Move logic of getInputsForSweep into getInputsForAmount for DRY

matthewjablack committed:
Add sendSweepTransaction to Ethereum ERC20 Provider

matthewjablack committed:
Update Chain client to use cross-chain sendSweepTransaction interface

matthewjablack committed:
Update ethereum js wallet provider sendSweepTransaction arg naming

matthewjablack commented:
@monokh added `sendSweepTransaction` to `EthereumERC20Provider` and addressed all your comments

monokh reviewed (changes requested):
This is looking in really good shape! Just one final comment

matthewjablack committed:
Remove Bitcoin specific input and output functionality for sweep

matthewjablack commented:
Updated @monokh to remove Bitcoin specific functionality. This can be implemented externally if developers wish to allow wasabi type functionality

monokh reviewed (approved):
Looking good! Thanks for this
